[pharrell.mp3 민필규] 지하철 노선도 step3 PR 요청 드립니다

pilgyumin opened the pull request:
안녕하세요 
pharrell(퍼렐_민필규)입니다.

이번 지하철 노선도는 이전 프로젝트보다 시간이 오래 걸렸던 것 같습니다.

자동차 경주, 로또보다 구현해야되는 알고리즘이 어려워서, 해당 부분을 가독성 있게
짜기 위해서 노력하였습니다.
그러다 보니 이전 프로젝트보다 시간이 오래 걸렸습니다.

지금 코드도 더 리팩토링해야될 것으로 느껴지지만, 시간을 더 이상 지체할 수는 없을 것 같아서
리뷰 요청드립니다.

사소한 피드백도 감사하게 받을 수 있으니, 아낌없이 피드백 주시면 감사하겠습니다.

그럼 이번 코드 리뷰도 잘 부탁드리겠습니다.

감사합니다.

pharrell-mp3 committed:
docs(readme.md) : 구현 기능 목록 작성

pharrell-mp3 committed:
feat(frontend) : frontend node build

pharrell-mp3 committed:
feat(subway) : 역 추가 기능 개발

- controller에 dao 단일 객체로 생성
- 동일 이름의 역 중복에 대한 예외 처리

pharrell-mp3 committed:
feat(subway) : 역 목록 조회 기능 개발

- 역 이름 배열과 역 목록 매핑 로직 구현

pharrell-mp3 committed:
feat(subway) : 역 삭제 기능 구현 및 구현 목록 update

- controller 내 역 삭제 기능 구현

pharrell-mp3 committed:
feat(line) : 노선 생성 기능 개발

- 노선 controller, dao 내 노선 생성 기능 개발
- 노선 객체 구현

pharrell-mp3 committed:
feat(line) : 노선 목록 조회 구현

pharrell-mp3 committed:
feat(line) : 노선 삭제 기능 구현

pharrell-mp3 committed:
feat(line) : 노선 정보 조회 구현

- 노선 정보를 아이디로 조회
- 조회된 노선 정보가 없을 시 예외 처리

pharrell-mp3 committed:
feat(line) : 노선 수정 구현

- 노선 내 이름 및 색 수정 구현

pharrell-mp3 committed:
docs(readme.md) : readme 포멧 수정

jypark5826 committed:
Merge pull request #1 from pilgyumin/step1

Step1

pharrell-mp3 committed:
docs(readme.md) : step2 기능 구현 목록 작성

pharrell-mp3 committed:
refactor(line,station) : dao 객체 싱글톤 패턴 적용

- dao 객체 생성 시 싱글톤 패턴으로 생성되게 구현

pharrell-mp3 committed:
feat(station) : controller와 service 로직 분리

- 비즈니스 로직을 StationService에 구현

pharrell-mp3 committed:
feat(line) : controller와 service 로직 분리

- 비즈니스 로직을 LineService에 구현

pharrell-mp3 committed:
feat(station, line) : 지하철 노선 추가 API 수정

- 상행, 하행, 구간 정보 등록

pharrell-mp3 committed:
docs(readme.md) : 기능 구현 목록 추가

- 구간(Section) DAO 정의 추가

pharrell-mp3 committed:
feat(section) : 구간 객체 및 DAO 정의

pharrell-mp3 committed:
feat(line) : DB구조에 맞는 객체 설계

- Section과 Line간의 관계 설정

pharrell-mp3 committed:
feat(line) : 구간 상행 종점 등록 기능 추가

- 구간 생성 시 상행 종점 등록 기능 구현

pharrell-mp3 committed:
feat(line) : 구간 하행 종점 등록 기능 추가

- 구간 생성 시 하행 종점 등록 기능 구현

pharrell-mp3 committed:
refactor(line) : 상행, 하행 종점 추가 기능 리팩토링

- map을 사용하여 구간 내 역 관계 확인 로직 추가

pharrell-mp3 committed:
feat(line) : 구간 추가 기능(갈래길 방지) 개발

- map을 사용하여 구간 내 갈래길 방지 기능 개발

pharrell-mp3 committed:
feat(line) : 구간 제거 기능 구현

- stationId, lineId를 통한 구간 제거 기능 구현

pharrell-mp3 committed:
feat(line) : 구간 추가에 대한 예외처리

- 역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 등록 불가 기능 구현
- SectionDistanceExceedException 정의

pharrell-mp3 committed:
feat(line) : 상행역, 하행역 둘 중 하나라도 포함이 되지 않으면 추가 불가

pharrell-mp3 committed:
feat(line) : 상행역, 하행역 둘 중 하나라도 포함이 되지 않으면 추가 불가

pharrell-mp3 committed:
feat(line) : 구간이 하나인 노선에서는 종점을 제거 불가

jypark5826 committed:
Merge pull request #2 from pilgyumin/step2

Step2

jamal-park committed:
docs(readme.md) : 구현 기능 목록 작성

jamal-park committed:
feat(spring) : 객체 스프링 빈으로 관리

jamal-park committed:
feat(station) : Dao JDBC 추가

pilgyumin committed:
Merge pull request #1 from jypark5826/jypark5826

Jypark5826

jamal-park committed:
feat(station) : JDBC 로직 수정

jamal-park committed:
feat(line) : LineDao JDBC 추가

jamal-park committed:
feat(section) : SectionDao JDBC 추가

jamal-park committed:
refactor(service) : Service 로직 리펙토링

pilgyumin committed:
Merge pull request #2 from jypark5826/jypark5826

Jypark5826

pharrell-mp3 committed:
refactor(station) : Station DB 연산에 사용되는 쿼리를 enum클래스에 정의

pharrell-mp3 committed:
refactor(line) : Section DB 연산에 사용되는 쿼리 enum 클래스에 정의

pharrell-mp3 committed:
refactor(line) : Line DB 연산에 사용되는 쿼리를 enum클래스에 정의

pharrell-mp3 committed:
refactor(line, section) : section에 관련된 클래스를 section 패키지로 이관

pharrell-mp3 committed:
refactor(line,section) : section 삭제에 대한 버그 수정

- 섹션 삭제 시에 라인에 해당 데이터 업데이트 부분 반영

pharrell-mp3 committed:
refactor(line,section) : section 병합 메소드 리팩토링

htrucci commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,51 @@
+@RestController
+public class LineController {
+
+    @Resource
+    private LineService lineService;
```
스프림프레임워크 환경에서는 주로 @Autowired를 사용하여 Bean을 주입합니다.
또한 개발자들이 초기에는 필드 주입방식을 많이 사용했는데 요즘은 생성자 주입 방식을 사용하는것이 추세인 것 같습니다.
아래 참고할 만한 내용 링크해드릴게요~!

[autowired 3가지 방법](https://junghyungil.tistory.com/50)
[생성자 주입을 해야하는 이유](https://yaboong.github.io/spring/2019/08/29/why-field-injection-is-bad/)

pilgyumin replied:
넵 해당 부분 반영하겠습니다!

htrucci commented on src/main/java/subway/line/LineController.java:
```diff
@@ -0,0 +1,51 @@
+    @Resource
+    private LineService lineService;
+
+    @PostMapping(value = "/lines")
+    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
+        return lineService.createLine(lineRequest);
+    }
+
+    @GetMapping(value = "/lines", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<List<LineResponse>> showLines() {
+        return lineService.showLines();
+    }
+
+    @DeleteMapping("/lines/{id}")
+    public ResponseEntity deleteLine(@PathVariable Long id) {
+        return lineService.deleteLine(id);
+    }
+
+    @GetMapping(value = "/lines/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity<LineResponse> showLine(@PathVariable Long id) {
+        return lineService.showLine(id);
+    }
+
+    @PutMapping(value = "/lines/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity updateLine(@PathVariable Long id, @RequestBody LineRequest lineRequest) {
+        return lineService.updateLine(id, lineRequest);
+    }
+
+    @PostMapping(value = "/lines/{id}/sections")
+    public ResponseEntity createSection(@PathVariable Long id, @RequestBody SectionRequest sectionRequest) {
+        return lineService.createSection(id, sectionRequest);
+    }
+
+    @DeleteMapping(value = "/lines/{id}/sections")
+    public ResponseEntity deleteSection(@PathVariable Long id, @RequestParam Long stationId) {
+        return lineService.deleteSection(id, stationId);
+    }
```
깔끔하네요 💯

LineController에서는 "/lines"를 모두 포함하는 공통점이 있으니 클래스 수준의 선언은 어떨까요?
@RequestMapping("/lines")

pilgyumin replied:
넵 해당 부분 반영하겠습니다!

htrucci commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,48 @@
+package subway.line;
```
현재 도메인별로 패키지가 되어있는데 아래와 같이 한뎁스 더 구조화한다면 좋을 것 같아요~!
lines.domain
lines.dao
lines.service
lines.controller

pilgyumin replied:
네 내부 패키지에 대해 구조화하겠습니다!

htrucci commented on src/main/java/subway/line/LineQuery.java:
```diff
@@ -0,0 +1,22 @@
+package subway.line;
+
+public enum LineQuery {
+    SAVE("insert into LINE (name, color, up_station_id, down_station_id, distance) values (?, ?, ?, ?, ?)"),
+    FIND_ALL( "select * from LINE"),
+    DELETE_BY_ID("delete from LINE where id = ?"),
+    FIND_BY_ID("select * from LINE where id = ?"),
+    FIND_BY_NAME("select * from LINE where name = ?"),
+    UPDATE("update LINE set up_station_id = ?, down_station_id = ?, distance = ? where id = ?")
+    ;
+
+
+    private LineQuery(String query) {
+        this.query = query;
+    }
+
+    public String getQuery() {
+        return query;
+    }
+
+    private String query;
+}
```
오~ 쿼리만 분리를 아주 잘 해주셨네요 💯

pilgyumin replied:
👍👍

htrucci commented on src/main/java/subway/line/LineDao.java:
```diff
@@ -0,0 +1,48 @@
+    @Resource
+    LineMapper lineMapper;
+
+    public void save(Line line) {
+        jdbcTemplate.update(LineQuery.SAVE.getQuery(),
+                line.getName(), line.getColor(), line.getUpStationId(), line.getDownStationId(), line.getDistance());
+    }
+
+    public List<Line> findAll() {
+        return jdbcTemplate.query(LineQuery.FIND_ALL.getQuery(), lineMapper);
+    }
+
+    public void deleteById(Long id) {
+        jdbcTemplate.update(LineQuery.DELETE_BY_ID.getQuery(),id);
+    }
+
+    public Line findById(Long id) {
+        return jdbcTemplate.queryForObject(LineQuery.FIND_BY_ID.getQuery(),lineMapper, id);
+    }
+
+    public Line findByName(String name) {
+        return jdbcTemplate.queryForObject(LineQuery.FIND_BY_NAME.getQuery(),lineMapper, name);
+    }
+
+    public void update(Line line) {
+        jdbcTemplate.update(LineQuery.UPDATE.getQuery(),
+                line.getUpStationId(), line.getDownStationId(), line.getDistance(), line.getId());
+    }
+}
```
잘해주셔서 딱히 코멘트 드릴 부분이 없네요

pilgyumin replied:
👍👍 감사합니다!

htrucci commented on src/main/java/subway/line/LineResponse.java:
```diff
@@ -9,15 +9,40 @@
         this.stations = stations;
     }
 
+    public LineResponse(Line line) {
+        this.id = line.getId();
+        this.name = line.getName();
+        this.color = line.getColor();
+    }
+
+
+    public LineResponse(Line line, List<Station> stations) {
+        if(!validator(line)){
+            throw new IllegalArgumentException("노선 정보가 존재하지 않습니다.");
+        }
+        this.id = line.getId();
+        this.name = line.getName();
+        this.color = line.getColor();
+        this.stations = stations;
+    }
+
```
💯

pilgyumin replied:
👍👍 감사합니다!

htrucci commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,292 @@
+    @Resource
+    private SectionDao sectionDao;
+
+    public ResponseEntity<LineResponse> createLine(LineRequest lineRequest) {
+        Line line = new Line(lineRequest);
+
+        try {
+            lineDao.save(line);
+        } catch (DuplicateKeyException dke){
+            return ResponseEntity.badRequest().build();
+        }
+
+        Line newLine = lineDao.findByName(line.getName());
+
+        sectionDao.save(new Section(newLine));
+
+        return ResponseEntity.created(URI.create("/lines/" + newLine.getId()))
+                .body(new LineResponse(newLine, getStations(newLine.getId())));
+    }
```
응답과 관련한 부분은 Controller에서 처리하는 것이 좋습니다.
따라서, 서비스에서는 해당 기능에 대한 결과만 리턴하고
ResponseEntity 처리와 관련한 부분은 컨트롤러에서 처리 부탁드려요~! 😄

pilgyumin replied:
넵 해당 부분 반영하겠습니다!

htrucci commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,292 @@
+
+    public ResponseEntity<List<LineResponse>> showLines() {
+        List<Line> lines = lineDao.findAll();
+        List<LineResponse> lineResponses = lines.stream().map(LineResponse::new).collect(Collectors.toList());
```
보기 쉽도록 개행을 해주는 것이 좋겠어요 😄

pilgyumin replied:
넵 코드 포멧팅 진행하겠습니다.

htrucci commented on src/main/java/subway/station/StationController.java:
```diff
@@ -4,27 +4,29 @@
 import java.util.List;
+import java.util.stream.Collectors;
 
 @RestController
 public class StationController {
```
여기도 클래스 수준에서 Url맵핑을 해주면 좋을 것 같네요~!
@RequestMapping("/stations")

pilgyumin replied:
네 이부분도 매핑 반영하겠습니다!

htrucci commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,36 @@
+    @Resource
+    public StationDao stationDao;
+
+    public ResponseEntity<StationResponse> createStation(StationRequest stationRequest) {
+        stationDao.save(new Station(stationRequest.getName()));
+
+        Station newStation = stationDao.findByName(stationRequest.getName());
+        StationResponse stationResponse = new StationResponse(newStation.getId(), newStation.getName());
+        return ResponseEntity.created(URI.create("/stations/" + newStation.getId())).body(stationResponse);
+    }
```
preparedStatement를 방식을 사용한다면 AutoGennerated된 key를 얻을 수 있습니다.
반환된 keyId로 findById를 한다면 더 명확한 로직이 될 거에요.


[preparedstatement with jdbctemplate](https://www.javatpoint.com/example-of-PreparedStatement-in-Spring-JdbcTemplate)

pilgyumin replied:
네 해당 부분 반영하겠습니다.

htrucci commented on src/main/java/subway/station/StationService.java:
```diff
@@ -0,0 +1,36 @@
+        return ResponseEntity.created(URI.create("/stations/" + newStation.getId())).body(stationResponse);
+    }
+
+    public ResponseEntity<List<StationResponse>> showStations() {
+        List<Station> stations = stationDao.findAll();
+        List<StationResponse> stationResponses = stations.stream().map(StationResponse::new).collect(Collectors.toList());
+        return ResponseEntity.ok().body(stationResponses);
+    }
+
+    public ResponseEntity deleteStation(Long id) {
+        stationDao.deleteById(id);
+        return ResponseEntity.noContent().build();
+    }
```
ResponseEntity처리와 관련된 부분은 Controller로 역할을 옮겨주세요~!

htrucci replied:
역 <-> 노선 <-> 구간 객체의 의존 및 데이터 종속성을 생각한다면
역이 삭제되는 경우 구간 데이터 정합이 깨질 것 같아요.
해당 역이 포함된 구간노선정보가 있다면? 삭제가 불가하게 하거나
역 삭제시 같이 지워지게 해야 서비스 안정성이 높아지겠네요! 😄

pilgyumin replied:
네 맞네요 해당 부분을 고려하지 못했었습니다.
해당 부분을 고려해서 다시 반영하겠습니다!

htrucci commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,292 @@
+                && stations.contains(stationDao.findById(sectionRequest.getDownStationId()));
+    }
+
+    public ResponseEntity deleteSection(Long id, Long stationId) {
+        //@TODO 삭제
+        // 1. stationId로 구간 조회
+        // 2. 해당 역이 상행, 하행 종점일 경우 구간 삭제 + 노선의 상행, 하행 업데이트 (sectionDao, LineDao)
+        // 3. 중간에 갈래길일 경우, 두개의 구간을 하나의 구간으로 통합
+        // 4. 역애 대한 정보 삭제 (stationDao)
+
+        List<Section> sections = sectionDao.findByLineId(id);
+        if (sections.size() == Line.END_STATION_SECTION_SIZE) {
+            throw new DeleteSectionException("구간이 하나인 노선에서 마지막 구간을 제거할 수 없음");
+        }
+
+        // 1.
+        List<Section> sectionList = sectionDao.findByStationIdAndLineId(stationId,id);
+
+        Line line = lineDao.findById(id);
+
+        // 2.
+        if(line.isEndStation(sectionList.size())){
+            Section endSection = sectionList.get(0);
+            line.updateEndStation(endSection,stationId);
+            sectionDao.deleteById(endSection.getId());
+        }
+
+        // 3.
+        if(!line.isEndStation(sectionList.size())){
+            sectionDao.deleteBySectionList(sectionList);
+            sectionDao.save(Section.merge(sectionList,stationId));
+        }
+
+        // 4.
+        stationDao.deleteById(stationId);
+        lineDao.update(line);
+
+        return ResponseEntity.ok().build();
+    }
```
3번과정에서 오류가 나면 어떻게 될까요?
구간삭제가 되었으나 역정보 삭제에 실패하고 데이터 정합성이 깨지게되겠죠.
따라서, 트랜잭션 개념을 적용해주는 것이 필요해졌습니다.
오류가 나지 않을것같거나 단순하더라도 입력, 수정, 삭제 로직은 서비스 레이어에서 @Transaction을 걸어주는게 좋습니다.

[@Transactional](https://goddaehee.tistory.com/167)

pilgyumin replied:
네 해당 부분 반영하겠습니다!

htrucci commented on src/main/java/subway/station/StationMapper.java:
```diff
@@ -0,0 +1,17 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component
+public class StationMapper implements RowMapper<Station>, Serializable {
+
+    @Override
+    public Station mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return new Station(rs.getLong("id"),rs.getString("name"));
+    }
+}
```
💯

pilgyumin replied:
ㅎㅎ 백점 감사합니다!!

htrucci commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,92 @@
+        return lineId;
+    }
+
+    public static Section merge(List<Section> sectionList, Long stationId) {
+        Section frontSection = sectionList.get(0);
+        Section backSection = sectionList.get(1);
+
+        if(frontSection.upStationId == stationId){
+            frontSection.upStationId = backSection.getUpStationId();
+        }
+
+        if(frontSection.downStationId == stationId){
+            frontSection.downStationId = backSection.getDownStationId();
+        }
+
+        frontSection.distance += backSection.getDistance();
+
+        return frontSection;
+    }
+
+    public static Map<Long, Section> getOrderedSections(List<Section> sections){
+        return sections.stream()
+                .collect(Collectors.toMap(Section::getUpStationId, section -> section));
+    }
+
+    public static Map<Long, Section> getReverseOrderedSections(List<Section> sections){
+        return sections.stream()
+                .collect(Collectors.toMap(Section::getDownStationId, section -> section));
+    }
```
List<Section> sections 를 Sections 일급컬렉션으로 만들고 관련 로직을 Sections쪽으로 분리해볼까요?

pilgyumin replied:
넵 Section List와 Station List를 일급컬렉션으로 정의하겠습니다.

htrucci commented on src/main/java/subway/line/LineMapper.java:
```diff
@@ -0,0 +1,23 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component
+public class LineMapper implements RowMapper<Line>, Serializable {
+
+    @Override
+    public Line mapRow(ResultSet rs, int rowNum) throws SQLException {
+        return new Line(rs.getLong("id"),
+                rs.getString("name"),
+                rs.getString("color"),
+                rs.getLong("up_station_id"),
+                rs.getLong("down_station_id"),
+                rs.getInt("distance"));
+    }
+}
```
💯

pilgyumin replied:
ㅎㅎㅎ 백점 감사합니다!!

htrucci commented on src/main/java/subway/line/LineResponse.java:
```diff
@@ -9,15 +9,40 @@
         this.stations = stations;
     }
 
+    public LineResponse(Line line) {
+        this.id = line.getId();
+        this.name = line.getName();
+        this.color = line.getColor();
+    }
```
생성자를 재활용할 수 있을 것 같네요.
this(line.getId(), line.getName(), line.getColor());

pilgyumin replied:
넵 해당 부분 반영하겠습니다!

htrucci commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,292 @@
+
+    public void addDownStation(Map<Long, Section> orderedSections, Line line, SectionRequest sectionRequest) {
+        Long upStationId = sectionRequest.getUpStationId();
+        int distanceSum = 0;
```
0도 의미를 가진 상수로 정의해볼까요! 👀

pilgyumin replied:
넵 SECTION_DISTANCE_SUM_DEFAULT 이 네이밍을 가진 상수로 정의하겠습니다!

htrucci commented on src/main/java/subway/line/LineService.java:
```diff
@@ -0,0 +1,292 @@
+            Section section = orderedSections.get(upStationId);
+
+            if (distanceSum + section.getDistance() == sectionRequest.getDistance()) {
+                throw new SectionDistanceExceedException("역 사이에 새로운 역을 등록할 경우 기존 역 사이 길이보다 크거나 같으면 등록을 할 수 없음");
```
같은 에러메시지가 반복사용되고 있네요~!

pilgyumin replied:
해당 에러 메세지는 exception의 기본 생성자에 반영하겠습니다.

htrucci reviewed (changes requested):
안녕하세요, 지하철 노선도 미션을 리뷰하게된 데빈입니다. 😄 
저번에도 느꼈었는데 역시나 코드를 깔끔하게 작성하는 능력이 매우 뛰어나신 것 같습니다. 👍 
기능이나 로직에 수정이 필요한 부분은 없으나 레이어 구조를 개선했으면 하는 부분이 있어 😭 
코멘트 남겨드렸으니 확인해주시고 다시 PR요청 부탁드립니다~! 😄

pharrell-mp3 committed:
refactor(line, station) : DI 방식을 생성자 주입 방식으로 변경

- Controller 내부의 service의 의존성을 생성자 주입 방식으로 반영

pharrell-mp3 committed:
refactor(line, station) : Url의 클래스 수준 선언 적용

- 공통되는 url을 RequestMapping을 이용하여 공통화

pharrell-mp3 committed:
refactor(line, station, section) : 각 도메인 별 내부 패키지 세분화

- 각 도메인 패키지 내에 domain, dao, service, controller 추가 패키지 적용

pharrell-mp3 committed:
refactor(line, station) : service에서의 리턴값 변경

- Service에서 ResponseEntity을 리턴하지 않고 기능에 대한 결과만 리턴

pharrell-mp3 committed:
refactor(line,station,section) : 코드 포멧팅 리팩토링

- 개행 수정, 코드 포멧팅, 불필요한 import 제거

pharrell-mp3 committed:
refactor(line, station, section) : DI 방식을 생성자 주입 방식으로 변경

- Dao, Service 내부의 객체 의존성을 생성자 주입 방식으로 반영

pharrell-mp3 committed:
refactor(line) : 생성자 재활용

- 기존에 정의되었던 생성자를 재활용하여 리팩토링

pharrell-mp3 committed:
refactor(line) : 섹션 거리의 합 상수 정의

- SECTION_DISTANCE_SUM_DEFAULT 이름을 가진 상수 정의

pharrell-mp3 committed:
refactor(section) : section list를 일급 컬렉션인 Sections로 정의

- List<Section> 을 래핑하는 Sections 컬렉션 정의

pharrell-mp3 committed:
refactor(station) : station list를 일급 컬렉션인 Stations로 정의

- List<Station> 을 래핑하는 Stations 컬렉션 정의

pharrell-mp3 committed:
refactor(exception) : exception 내 공통되는 에러 메세지를 기본 생성자에 적용

pharrell-mp3 committed:
refactor(station) : save 연산 반환형 수정

- stationId을 받아 사용할 수 있도록 save 연산 반환형을 Long으로 설정

pharrell-mp3 committed:
refactor(section) : save 연산 반환형 수정

- sectionId을 받아 사용할 수 있도록 save 연산 반환형을 Long으로 설정

pharrell-mp3 committed:
refactor(line) : save 연산 반환형 수정

- lineId을 받아 사용할 수 있도록 save 연산 반환형을 Long으로 설정

pharrell-mp3 committed:
refactor(section) : 역 정보 삭제시 구간 정보 삭제 반영

- 역 삭제 시 역을 가지고 있는 구간 정보 조회 후 해당 구간 삭제

pharrell-mp3 committed:
refactor(line) : updateSectionList 변수명을 updateSections로 변경

pharrell-mp3 committed:
refactor(line,station) : transaction을 고려한 어노테이션 반영

- db 연산 수행시 데이터 정합성을 유지하기 위한 transactional 어노테이션 설정

pharrell-mp3 committed:
refactor(exception, line, section, station) : 코드 포멧팅

- 코드 포멧팅, 불필요한 변수 제거, 개행 정리

pharrell-mp3 committed:
feat(exception) : controllerAdvice 추가

- controller 내부에 전역으로 exception을 잡을 수 있는 controllerAdvice 및 exceptionHandler 추가

htrucci commented on src/main/java/subway/station/dao/StationDao.java:
```diff
@@ -0,0 +1,50 @@
+                stationMapper, id);
+    }
+
+    public void deleteById(Long id) {
+        jdbcTemplate.update(StationQuery.DELETE_BY_ID.getQuery(), id);
+    }
```
어플리케이션 요건상 리턴값 필요하지 않아서 없어도 되긴한데
때에 따라 쿼리 수행여부에 따라 다음 로직 진행여부가 갈리기도 하고 
쿼리 수행 건수가 필요할 수도 있어서 저는 가능한면 그냥 바로 return합니다. 😄 
(바꾸라고 말씀드린 건 아니고 저희 경우 습관적으로 그렇게 한다는 얘기입니당..  ㅋㅋ)

pilgyumin replied:
넵 알겠습니다. 해당 부분 인지하겠습니다!

htrucci commented on src/main/java/subway/station/service/StationService.java:
```diff
@@ -0,0 +1,65 @@
+    public void deleteStation(Long stationId) {
+        Sections sections = new Sections(sectionDao.findByLineId(stationId));
+
+        if (sections.size() == Line.END_STATION_SECTION_SIZE) {
+            throw new DeleteSectionException("구간이 하나인 노선에서 마지막 구간을 제거할 수 없음");
+        }
+
+        if (sections.size() == EMPTY_STATION_SIZE) {
+            return;
+        }
```
size를 가져와서 비교하지 않고 Sections에게 물어보는 방식은 어떨까요?
sections.is최소구간사이즈()
sections.isEmpty()

pilgyumin replied:
넵 sections에 메세지를 보내는 방식으로 변경하겠습니다!

htrucci commented on src/main/java/subway/station/dao/StationQuery.java:
```diff
@@ -0,0 +1,21 @@
+
+public enum StationQuery {
+    SAVE("insert into STATION (name) values (?)"),
+    FIND_ALL("select * from STATION"),
+    FIND_BY_ID("select * from STATION where id = ?"),
+    FIND_BY_NAME("select * from STATION where name = ?"),
```
아스타(*)를 사용하는과 컬럼명을 명시하는 것 중에서
성능이 좀 더 좋다고 컬렴명을 명시해야한다고 하는 얘기도 있습니다만,,
개인적으로는 필요한 컬럼 데이터만 가져오는 것이 더 효율적이기도 하고 DB테이블 스키마 변경에서도 
추적에 용이하기 떄문에 좀 더 안전하다고 생각합니다.
요즘은 트렌드인 JPA라는 것을 사용하여 쿼리를 작성할일이 많지 않고 
Mapper없이 객체와 테이블을 자동으로 맵핑해주기 때문에 
위와 같은 내용을 고민할일이 크게 없습니다. 😄 

[JPA란](https://blog.woniper.net/255)

pilgyumin replied:
넵 알겠습니다 좀 더 효율적인 쿼리 작성을 위해 컬럼명을 명시하겠습니다!
그리고 JPA에 관련된 글도 읽어보겠습니다!

htrucci commented on src/main/java/subway/station/service/StationService.java:
```diff
@@ -0,0 +1,65 @@
+        this.lineService = lineService;
+    }
+
+    @Transactional
+    public StationResponse createStation(StationRequest stationRequest) {
+        Long stationId = stationDao.save(new Station(stationRequest.getName()));
+        Station newStation = stationDao.findById(stationId);
+
+        return new StationResponse(stationId, newStation.getName());
+    }
```
피드백 잘 반영해주셨네요 😄 
다시 생각해보면, id와 name을 모두 알기때문에 DB질의를 하지 않아도 됩니다.
웹어플리케이션에서 응답시간에 많은 영향을 끼치는 부분이 DB에 갔다오는 부분입니다.
따라서 웹 어플리케이션에서는 가능한 DB질의 횟수를 줄이는게 중요합니다.

pilgyumin replied:
아 맞네요! stationRequest에서 name에 대한 정보를 가지고 올 수 있네요!!
해당 부분 수정하겠습니다!

htrucci commented on src/main/java/subway/line/domain/LineResponse.java:
```diff
@@ -0,0 +1,75 @@
+    private String name;
+    private String color;
+    private int extraFare;
+    private Stations stations;
+
+    public LineResponse() {
+    }
+
+    public LineResponse(Long id, String name, String color) {
+        this.id = id;
+        this.name = name;
+        this.color = color;
+    }
+
+    public LineResponse(Long id, String name, String color, Stations stations) {
+        this.id = id;
+        this.name = name;
+        this.color = color;
+        this.stations = stations;
+    }
```
Stations로 잘 랩핑을 해주셨는데 이로 인해 리턴되는 Json이 변경되어 오류가 나는 것 같네요.
프론트를 수정하지 않는다면 랩핑클래스는 그대로 사용해주시되 
LineResponse의 필드는 List<Station> stations; 를 유지해야겠어요. 😢

pilgyumin replied:
넵 LineResponse은 List<Station> stations를 내부 필드로 가지도록 변경하겠습니다!

htrucci commented on src/main/java/subway/line/domain/LineResponse.java:
```diff
@@ -0,0 +1,75 @@
+import subway.station.domain.Stations;
+
+public class LineResponse {
+    private int status;
```
사용하지 않는 필드로 보입니다!

pilgyumin replied:
넵 해당 필드 삭제하겠습니다!

htrucci commented on src/main/java/subway/line/service/LineService.java:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        if (!containsEndStation(id, sectionRequest)) {
+            throw new IllegalArgumentException("상행역과 하행역 둘 중 하나도 포함되어있지 않으면 추가할 수 없음");
```
반복 사용되는 메시지입니다.

pilgyumin replied:
넵 에러 메세지를 상수화 하겠습니다!

htrucci reviewed (approved):
안녕하세요 데빈입니다.
커밋 히스토리 갯수가 엄청나네요! 하나씩 따라가며 변화되는 과정을 보는 재미가 있었습니다. 👍 
이 PR은 머지를 하도록 하고 버그가 하나 있어 이 부분은 다음단계 진행하며 고쳐주세요 ㅎ
